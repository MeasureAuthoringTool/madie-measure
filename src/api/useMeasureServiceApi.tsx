import React from "react";
import axios from "axios";
import useServiceConfig from "./useServiceConfig";
import { ServiceConfig } from "./ServiceContext";
import { Measure, Group, Organization } from "@madie/madie-models";
import { useOktaTokens } from "@madie/madie-util";
import _ from "lodash";

export class MeasureServiceApi {
  constructor(private baseUrl: string, private getAccessToken: () => string) {}

  async fetchMeasure(id: string): Promise<Measure> {
    try {
      const response = await axios.get<Measure>(
        `${this.baseUrl}/measures/${id}`,
        {
          headers: {
            Authorization: `Bearer ${this.getAccessToken()}`,
          },
        }
      );
      return response.data;
    } catch (err) {
      const message = `Unable to fetch measure ${id}`;
      console.error(message);
      console.error(err);
      throw new Error(err);
    }
  }

  async fetchMeasures(
    filterByCurrentUser: boolean,
    limit: number = 25,
    page: number = 0
  ): Promise<any> {
    try {
      const response = await axios.get<any>(`${this.baseUrl}/measures`, {
        headers: {
          Authorization: `Bearer ${this.getAccessToken()}`,
        },
        params: {
          currentUser: filterByCurrentUser,
          limit,
          page,
        },
      });
      return response.data;
    } catch (err) {
      const message = `Unable to fetch measures`;
      console.error(message);
      console.error(err);
      throw new Error(message);
    }
  }

  async updateMeasure(measure: Measure): Promise<Response> {
    return await axios.put(`${this.baseUrl}/measures/${measure.id}`, measure, {
      headers: {
        Authorization: `Bearer ${this.getAccessToken()}`,
      },
    });
  }

  async createGroup(group: Group, measureId: string): Promise<Group> {
    try {
      const response = await axios.post<Group>(
        `${this.baseUrl}/measures/${measureId}/groups/`,
        group,
        {
          headers: {
            Authorization: `Bearer ${this.getAccessToken()}`,
          },
        }
      );
      if (response?.data) {
        const responseData = JSON.stringify(response.data);
        if (responseData.includes("Request Rejected")) {
          throw new Error("failure create group");
        }
      }
      return response.data;
    } catch (err) {
      const message = this.buildErrorMessage(
        err,
        "Failed to create the group."
      );
      console.error(message, err);
      throw new Error(message);
    }
  }

  async deleteMeasureGroup(
    groupId: string,
    measureId: string
  ): Promise<Measure> {
    try {
      if (groupId && measureId) {
        const response = await axios.delete(
          `${this.baseUrl}/measures/${measureId}/groups/${groupId}`,
          {
            headers: {
              Authorization: `Bearer ${this.getAccessToken()}`,
            },
          }
        );
        return response.data;
      } else {
        console.error("Group or Measure Id cannot be null");
        throw new Error("Group or Measure Id cannot be null");
      }
    } catch (err) {
      const message = this.buildErrorMessage(
        err,
        "Failed to delete the measure group."
      );
      console.error(message, err);
      throw new Error(message);
    }
  }

  async updateGroup(group: Group, measureId: string): Promise<Group> {
    try {
      const response = await axios.put<Group>(
        `${this.baseUrl}/measures/${measureId}/groups/`,
        group,
        {
          headers: {
            Authorization: `Bearer ${this.getAccessToken()}`,
          },
        }
      );
      if (response?.data) {
        const responseData = JSON.stringify(response.data);
        if (responseData.includes("Request Rejected")) {
          throw new Error("failure update group");
        }
      }
      return response.data;
    } catch (err) {
      const message = this.buildErrorMessage(
        err,
        "Failed to update the group."
      );
      console.error(message, err);
      throw new Error(message);
    }
  }

  buildErrorMessage(err, baseMessage): string {
    let errorMessage = undefined;
    if (err?.response?.status === 400) {
      if (err.response.data?.validationErrors?.group) {
        errorMessage =
          "Missing required populations for selected scoring type.";
      } else if (err.response.data.message) {
        errorMessage = err.response.data.message;
      }
    }
    return errorMessage ? errorMessage : baseMessage;
  }

  async getAllPopulationBasisOptions(): Promise<string[]> {
    try {
      const response = await axios.get<string[]>(
        `${this.baseUrl}/populationBasisValues`,
        {
          headers: {
            Authorization: `Bearer ${this.getAccessToken()}`,
          },
        }
      );
      if (response?.data.length < 1) {
        throw new Error("Unable to fetch population basis options");
      }
      return response?.data;
    } catch (err) {
      const message = this.buildErrorMessage(
        err,
        "Unable to fetch population basis options"
      );
      throw new Error(message);
    }
  }

  async getAllOrganizations(): Promise<Organization[]> {
    try {
      const response = await axios.get<Organization[]>(
        `${this.baseUrl}/organizations`,
        {
          headers: {
            Authorization: `Bearer ${this.getAccessToken()}`,
          },
        }
      );
      if (response?.data.length < 1) {
        throw new Error("Unable to fetch organizations");
      }
      return response?.data;
    } catch (err) {
      const message = this.buildErrorMessage(
        err,
        "Unable to fetch organizations"
      );
      throw new Error(message);
    }
  }

  getReturnTypesForAllCqlDefinitions(elmJson: string): {
    [key: string]: string;
  } {
    if (!elmJson) {
      return {};
    }

    const elm = JSON.parse(elmJson);
    const definitions = elm.library?.statements?.def;
    if (definitions) {
      return definitions.reduce((returnTypes, definition) => {
        const returnType = {};
        const name = _.camelCase(_.trim(definition.name));
        if (definition.resultTypeName) {
          returnType[name] = definition.resultTypeName?.split("}")[1];
        } else if (definition.resultTypeSpecifier) {
          for (const key in definition.resultTypeSpecifier) {
            if (
              definition.resultTypeSpecifier[key]?.type === "NamedTypeSpecifier"
            ) {
              returnType[name] =
                definition.resultTypeSpecifier[key].name?.split("}")[1];
            }
          }
        }
        // default to NA
        if (!returnType[name]) {
          returnType[name] = "NA";
        }
        return {
          ...returnTypes,
          ...returnType,
        };
      }, {});
    }

    return {};
  }

  getReturnTypesForAllCqlFunctions(elmJson: string): {
    [key: string]: string;
  } {
    if (!elmJson) {
      return {};
    }

    const elm = JSON.parse(elmJson);
    const functions = elm.library?.statements?.def?.filter(
      (func) => func.type === "FunctionDef"
    );

    if (functions) {
      return functions.reduce((returnTypes, func) => {
        const returnType = {};
        const name = _.camelCase(_.trim(func.name));
        if (
          func?.operand.length === 1 &&
          _.camelCase(
            func.operand[0].operandTypeSpecifier?.name?.split("}")[1]
          ) !== "boolean"
        ) {
          const operandSpecifier = func.operand[0].operandTypeSpecifier;
          // argument type that we are checking must match population basis
          // discarding list and interval as valid because we don't have any suitable value for them in population basis
          if (
            operandSpecifier?.name &&
            operandSpecifier?.type === "NamedTypeSpecifier"
          ) {
            returnType[name] = operandSpecifier?.name?.split("}")[1];
          } else {
            returnType[name] = "N/A";
          }
        } else if (func?.operand.length < 1) {
          returnType[name] = "Boolean";
        } else {
          returnType[name] = "N/A";
        }
        return {
          ...returnTypes,
          ...returnType,
        };
      }, {});
    }

    return {};
  }

  async searchMeasuresByMeasureNameOrEcqmTitle(
    filterByCurrentUser: boolean,
    limit: number = 25,
    page: number = 0,
    searchCriteria: string
  ): Promise<any> {
    try {
      const response = await axios.get<any>(
        `${this.baseUrl}/measures/search/${encodeURI(searchCriteria)}`,
        {
          headers: {
            Authorization: `Bearer ${this.getAccessToken()}`,
          },
          params: {
            currentUser: filterByCurrentUser,
            limit,
            page,
          },
        }
      );
      return response.data;
    } catch (err) {
      const message = `Unable to search measures`;
      console.error(message);
      console.error(err);
      throw new Error(message);
    }
  }

  async createVersion(id: string, versionType: string): Promise<Measure> {
    return await axios.put(
      `${this.baseUrl}/measures/${id}/version/?versionType=${versionType}`,
      {},
      {
        headers: {
          Authorization: `Bearer ${this.getAccessToken()}`,
        },
      }
    );
  }

  async draftMeasure(measureId: string, measureName: string): Promise<Measure> {
    return await axios.post(
      `${this.baseUrl}/measures/${measureId}/draft`,
      { measureName: measureName },
      {
        headers: {
          Authorization: `Bearer ${this.getAccessToken()}`,
        },
      }
    );
  }
}

export default function useMeasureServiceApi(): MeasureServiceApi {
  const serviceConfig: ServiceConfig = useServiceConfig();
  const { getAccessToken } = useOktaTokens();
  const { baseUrl } = serviceConfig.measureService;

  return new MeasureServiceApi(baseUrl, getAccessToken);
}
