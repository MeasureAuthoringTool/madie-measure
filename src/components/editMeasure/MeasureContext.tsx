import { createContext } from "react";
import { Measure } from "../../../../madie-models/dist";

export interface MeasureContextHolder {
  measure: Measure;
  setMeasure: (measure: Measure) => void;
}

const MeasureContext = createContext<MeasureContextHolder>(null);

export default MeasureContext;

export const MeasureContextProvider = MeasureContext.Provider;
export const MeasureContextConsumer = MeasureContext.Consumer;
