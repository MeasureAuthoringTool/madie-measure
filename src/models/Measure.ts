import { Model } from "./Model";
import { MeasureScoring } from "./MeasureScoring";

export interface MeasureMetadata {
  measureSteward?: string;
  measureDescription?: string;
}

export default interface Measure {
  id: string;
  measureHumanReadableId: string;
  measureSetId: string;
  version: number;
  revisionNumber: number;
  state: string;
  measureName: string;
  cqlLibraryName: string;
  measureScoring: MeasureScoring | "";
  cql: string;
  createdAt: string;
  createdBy: string;
  lastModifiedAt: string;
  lastModifiedBy: string;
  model: Model | "";
  measureMetaData?: MeasureMetadata;
}
