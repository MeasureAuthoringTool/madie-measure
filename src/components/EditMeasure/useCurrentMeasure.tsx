import React, { useContext } from "react";
import MeasureContext from "./MeasureContext";
import Measure from "../../models/Measure";

export default function useCurrentMeasure() {
  const measure: Measure = useContext(MeasureContext);
  return measure;
}
