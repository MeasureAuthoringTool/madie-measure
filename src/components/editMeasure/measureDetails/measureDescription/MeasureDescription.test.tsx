import React from "react";
import { render, cleanup, fireEvent } from "@testing-library/react";
import MeasureDescription from "./MeasureDescription";
import useMeasureServiceApi, {
  MeasureServiceApi,
} from "../../../../api/useMeasureServiceApi";
import useCurrentMeasure from "../../useCurrentMeasure";
import { MeasureContextHolder } from "../../MeasureContext";
import Measure from "../../../../models/Measure";

jest.mock("../../../../api/useMeasureServiceApi");
jest.mock("../../useCurrentMeasure");

const useMeasureServiceApiMock =
  useMeasureServiceApi as jest.Mock<MeasureServiceApi>;

const useCurrentMeasureMock =
  useCurrentMeasure as jest.Mock<MeasureContextHolder>;

const DESCRIPTION = "Test Description";

describe("MeasureDescription component", () => {
  let measure: Measure;
  let serviceApiMock: MeasureServiceApi;
  let measureContextHolder: MeasureContextHolder;

  afterEach(cleanup);

  beforeEach(() => {
    measure = {
      id: "test measure",
      measureName: "the measure for testing",
      measureMetaData: {
        measureDescription: DESCRIPTION,
      },
    } as Measure;

    serviceApiMock = {
      updateMeasure: jest.fn().mockResolvedValue(undefined),
    } as unknown as MeasureServiceApi;

    useMeasureServiceApiMock.mockImplementation(() => serviceApiMock);

    measureContextHolder = {
      measure,
      setMeasure: jest.fn(),
    };

    useCurrentMeasureMock.mockImplementation(() => measureContextHolder);
  });

  const expectInputValue = (element: HTMLElement, value: string): void => {
    expect(element).toBeInstanceOf(HTMLInputElement);
    const inputEl = element as HTMLInputElement;
    expect(inputEl.value).toBe(value);
  };

  it("should render the component with the supplied description information", () => {
    const { getByTestId } = render(<MeasureDescription />);

    const result = getByTestId("measureDescription");
    expect(result).toBeInTheDocument();
    expect(result).toMatchSnapshot();

    const input = getByTestId("measureDescriptionInput");
    expectInputValue(input, DESCRIPTION);
  });

  it("should default the measureMetadata if none is supplied", async () => {
    delete measure.measureMetaData;
    const { findByTestId, getByTestId } = render(<MeasureDescription />);
    const input = getByTestId("measureDescriptionInput");
    expectInputValue(input, "");

    const save = getByTestId("measureDescriptionSave");
    fireEvent.click(save);

    expect(serviceApiMock.updateMeasure).toHaveBeenCalledWith({
      id: "test measure",
      measureMetaData: {
        measureDescription: undefined,
      },
      measureName: "the measure for testing",
    });

    const success = await findByTestId("measureDescriptionSuccess");
    expect(success).toBeInTheDocument();
  });

  it("should update the input when a user types a new value", () => {
    const { getByTestId } = render(<MeasureDescription />);

    const input = getByTestId("measureDescriptionInput");
    expectInputValue(input, DESCRIPTION);

    fireEvent.change(input, {
      target: { value: "new value" },
    });

    expectInputValue(input, "new value");
  });

  it("should save the description information when the form is submitted", async () => {
    const { findByTestId, getByTestId } = render(<MeasureDescription />);
    const input = getByTestId("measureDescriptionInput");
    fireEvent.change(input, {
      target: { value: "new value" },
    });
    const save = getByTestId("measureDescriptionSave");
    fireEvent.click(save);

    expect(serviceApiMock.updateMeasure).toHaveBeenCalledWith({
      id: "test measure",
      measureMetaData: {
        measureDescription: "new value",
      },
      measureName: "the measure for testing",
    });

    const success = await findByTestId("measureDescriptionSuccess");
    expect(success).toBeInTheDocument();
  });

  it("should render an error message if the measure cannot be saved", async () => {
    serviceApiMock.updateMeasure = jest.fn().mockRejectedValue("Save error");

    const { findByTestId, getByTestId } = render(<MeasureDescription />);
    const input = getByTestId("measureDescriptionInput");
    fireEvent.change(input, {
      target: { value: "new value" },
    });
    const save = getByTestId("measureDescriptionSave");
    fireEvent.click(save);

    const error = await findByTestId("measureDescriptionError");
    expect(error.textContent).toBe(
      'Error updating measure "the measure for testing"'
    );
  });
});
