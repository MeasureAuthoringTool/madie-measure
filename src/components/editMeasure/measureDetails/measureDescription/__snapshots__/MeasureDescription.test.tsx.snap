// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MeasureDescription component should render the component with the supplied description information 1`] = `
<form
  class="sc-bdvvtL exmvhH"
  data-testid="measureDescription"
>
  <div
    class="sc-gsDKAQ jTbVDJ"
  >
    <div>
      <div
        class="sc-dkPtRN bvfDZg"
      >
        <div
          class="sc-hKwDye ZBVeD"
        >
          <label
            class="sc-eCImPb dMhjhF"
            for="measure-description"
          >
            Measure Description
          </label>
          <div
            class="sc-jRQBWg dA-DgiC"
          >
            <input
              autocomplete="given-name"
              class="sc-gKclnd jEPGhr"
              data-testid="measureDescriptionInput"
              id="measure-description"
              name="measure-description"
              placeholder="Measure Description"
              required=""
              type="text"
              value="Test Description"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="sc-iCfMLu Gwmnn"
  >
    <div
      class="sc-furwcr dKjrjs"
    >
      <button
        class="sc-pVTFL cSoTMs"
        data-testid="measureDescriptionSave"
        type="submit"
      >
        Save
      </button>
      <div
        class="sc-jrQzAO wImVQ"
      >
        
      </div>
    </div>
  </div>
</form>
`;
